File: GMUtilsFlutter/gmutils_flutter/lib/zgmutils/data_utils/firebase/firebase_utils.dart
// import 'package:http/http.dart' as http;

class FirebaseUtils {
  //region refineKeyName
  static String refineKeyName(String name) => name
      .trim()
      .replaceAll(' ', '_')
      .replaceAll('.', '_')
      .replaceAll('\$', '_')
      .replaceAll('#', '_')
      .replaceAll('@', '_')
      .replaceAll('[', '_')
      .replaceAll(']', '_')
      .replaceAll('/', '_');

  //endregion

  //region refinePath
  static String refinePathFragmentNames(
    String path, {
    bool excludeLastDot = false,
  }) {
    var ext = '';
    if (excludeLastDot) {
      var i = path.lastIndexOf('.');
      if (i >= 0 && i < path.length - 1) {
        ext = '.' + path.substring(i + 1);
        path = path.substring(0, i);
      }
    }

    var frags = path.split("/");

    path = '';
    for (var value in frags) {
      value = refineKeyName(value);
      if (value.isEmpty) throw "invalid_node_name";

      if (path.isNotEmpty) path += '/';
      path += value;
    }

    return path + ext;
  }

  //endregion

  //region refinePhoneNumber
  static String refinePhoneNumber(String number) => number
      .trim()
      .replaceAll('-', '')
      .replaceAll(' ', '')
      .replaceAll('.', '')
      .replaceAll('+', '')
      .replaceAll('(', '')
      .replaceAll('/', '')
      .replaceAll(')', '')
      .replaceAll('N', '')
      .replaceAll(',', '')
      .replaceAll('*', '')
      .replaceAll(';', '')
      .replaceAll('#', '');

//endregion

//region isConnectionAvailable
/*static int _lastConnectionCheckTime = 0;
  static bool _lastConnectionCheckResult = false;

  static Future<bool> isConnectionAvailable() async {
    try {
      int now = DateTime.now().millisecondsSinceEpoch;
      int diff = now - _lastConnectionCheckTime;
      _lastConnectionCheckTime = now;
      if (diff < 10000 && _lastConnectionCheckResult) return true;

      var response = await http.get(Uri.parse('https://www.google.com'));
      _lastConnectionCheckResult = true;
      Logs.print(() => 'FirebaseUtils.isConnectionAvailable ---> response of: google.com: ${response.statusCode}');
      return true;
    } catch (e) {
      _lastConnectionCheckResult = false;
      return false;
    }
  }*/
//endregion
}

File: GMUtilsAndroid/GMUtils/GMUtils/src/main/java/gmutils/firebase/FirebaseUtils.java
package gmutils.firebase;

import gmutils.listeners.ResultCallback;
import gmutils.net.SimpleHTTPRequest;

public class FirebaseUtils {
    //region refineKeyName
    public static String refineKeyName(String name) {
        return name
                .trim()
                .replace(' ', '_')
                .replace('.', '_')
                .replace('$', '_')
                .replace('#', '_')
                .replace('@', '_')
                .replace('[', '_')
                .replace(']', '_')
                .replace('/', '_');
    }

    //endregion

    //region refinePath
    public static String refinePathFragmentNames(String path) throws Exception {
        var frags = path.split("/");//a,b
        StringBuilder pathBuilder = new StringBuilder();
        for (String value : frags) {
            value = refineKeyName(value);
            if (value.isEmpty()) throw new Exception("invalid_node_name");

            if (!pathBuilder.toString().isEmpty()) pathBuilder.append("/");
            pathBuilder.append(value);
        }
        path = pathBuilder.toString();
        return path;
    }

    //endregion

    //region refinePhoneNumber
    public static String refinePhoneNumber(String number) {
        return number
                .trim()
                .replace("-", "")
                .replace(" ", "")
                .replace(".", "")
                .replace("+", "")
                .replace("(", "")
                .replace("/", "")
                .replace(")", "")
                .replace("N", "")
                .replace(",", "")
                .replace("*", "")
                .replace(";", "")
                .replace("#", "");

    }

    //endregion

    //region isConnectionAvailable
    private static long _lastConnectionCheckTime = 0;
    private static boolean _lastConnectionCheckResult = false;

    public static void isConnectionAvailable(ResultCallback<Boolean> callback) {
        try {
            long now = System.currentTimeMillis();
            long diff = now - _lastConnectionCheckTime;
            _lastConnectionCheckTime = now;
            if (diff < 15000 && _lastConnectionCheckResult) callback.invoke(true);

            SimpleHTTPRequest.call(
                    true,
                    new SimpleHTTPRequest.Request(
                            "https://www.google.com",
                            SimpleHTTPRequest.Method.GET
                    ),
                    null,
                    (result1, result2) -> {
                        callback.invoke(result2.getException() == null);
                        _lastConnectionCheckResult = true;
                    }
            );

        } catch (Exception e) {
            _lastConnectionCheckResult = false;
            callback.invoke(false);
        }
    }
    //endregion
}
